### PR TITLE
Practice with cucumber and rest Assured

### PR DESCRIPTION
**Pull request by rosariogarcia**
Hi @Carledriss I send you my progress in the task

### CODE REVIEW

File: pivotal-api-test/src/test/resources/project/CreateProject [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Feature: Create new project in pivotal tracker
```
Comment (carledriss) [inline pivotal-api-test/src/test/resources/project/CreateProject]: Feature files should have the .feature extension

File: pivotal-api-test/src/test/resources/stories/UpdateStory [hunk 2]
```diff
@@ -0,0 +1,19 @@
+Feature: Update a story in project from pivotal tracker
```
Comment (carledriss) [inline pivotal-api-test/src/test/resources/stories/UpdateStory]: Notice that the Update and Delete story have the same preconditions (Background)
Probably you can join them in a single feature file.

File: pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/RequestStepDef.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import cucumber.api.java.en.Given;
+
+/**
+ * Created by RosarioGarcia on 6/30/2016.
+ */
+public class RequestStepDef {
+    private Response response;
+
+    @Given("^I send a POST request to (.*) with:$")
+    public void iSendAPOSTRequestToProjects(String endpoint, Map<String, Object> values) {
```
Comment (carledriss) [inline pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/RequestStepDef.java]: Method name is too specific. Rename it

File: pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/RequestStepDef.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        response = RequestManager.postRequest(LocalStore.formatEndpoint(endpoint), values);
+    }
+
+    @Given("^I send a PUT request to (.*) with:$")
+    public void iSendAPUTRequestToProjectsIdWith(String endpoint, Map<String, Object> values) {
```
Comment (carledriss) [inline pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/RequestStepDef.java]: same as above

File: pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/RequestStepDef.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        response = RequestManager.putRequest(LocalStore.formatEndpoint(endpoint), values);
+    }
+
+    @Given("^I send a DELETE request to (.*)$")
+    public void iSendADELETERequestToProjects(String endpoint) {
```
Comment (carledriss) [inline pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/RequestStepDef.java]: same as above

File: pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/Hook.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+public class Hook {
+    Response response;
+
+    public Hook(RequestStepDef requestStepDef) {
+        this.response = requestStepDef.getResponse();
+    }
+
+    @After("@deleteProject")
+    public void tearDownProject() {
+        response = RequestManager.deleteRequest("/projects/" + response.jsonPath().get("id"));
```
Comment (carledriss) [inline pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/Hook.java]: Create a constant for "/projects/"

File: pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/LocalStore.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+ */
+public class LocalStore {
+    private static Map<String, Response> mapResponse;
+
+    private LocalStore() {
+    }
+
+    public static String formatEndpoint(String endpoint) {
+        if (endpoint.contains("[")) {
+            Pattern keyEndpoint = Pattern.compile("\\[(.*?)\\.");
```
Comment (carledriss) [inline pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/LocalStore.java]: create constants for regex

File: pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/LocalStore.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+            Matcher mKey = keyEndpoint.matcher(endpoint);
+            Pattern valueEndpoint = Pattern.compile("\\.(.*?)\\]");
```
Comment (carledriss) [inline pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/LocalStore.java]: same as above

File: pivotal-api-test/pivotal.properties [hunk 9]
```diff
@@ -0,0 +1,3 @@
+urlApi = https://www.pivotaltracker.com/services/v5/
```
Comment (carledriss) [inline pivotal-api-test/pivotal.properties]: Remove the last /
It should be 

```
urlApi = https://www.pivotaltracker.com/services/v5
```

**Comment by rosariogarcia** [comment]
@carledriss, I upload the fix to your last observations.

File: pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/Authentication.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * This Singleton class is for connect with the Pivotal Tracker API
+ * with valid credentials, in this case use the Token of user.
+ *
+ * @author RosarioGarcia
+ */
+public final class Authentication {
+
+    private static final String TYPE_HEADER = "X-TrackerToken";
```
Comment (carledriss) [inline pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/Authentication.java]: I would prefer to call it TOKEN_HEADER

File: pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/Authentication.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * Getter to return the requestSpecification variable
+     * that contains the request to connect to pivotal API
+     *
+     * @return requestSpecification
+     */
+    public RequestSpecification getRequestSpecification() {
+        return requestSpecification;
+    }
+
```
Comment (carledriss) [inline pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/Authentication.java]: Remove this unnecessary line

File: pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/LocalStore.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
+package com.fundacionjala.pivotalapi;
+
+import com.jayway.restassured.response.Response;
```
Comment (carledriss) [inline pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/LocalStore.java]: Organize the imports

File: pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/Authentication.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+}
```
Comment (carledriss) [inline pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/Authentication.java]: Add a  blank space at the end of the class

File: pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/LocalStore.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+ * In this class build the Map of responses
+ * from requests to Pivotal API.
+ *
+ * @author RosarioGarcia
+ */
+public class LocalStore {
+    private static final Logger LOGGER = Logger.getLogger(PropertiesInfo.class);
+    public static final String REGEX_KEY = "\\[(.*?)\\.";
```
Comment (carledriss) [inline pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/LocalStore.java]: this should be private

File: pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/LocalStore.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    public static final String REGEX_VALUE = "\\.(.*?)\\]";
```
Comment (carledriss) [inline pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/LocalStore.java]: same as above

File: pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/LocalStore.java [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    public static final String REGEX_REPLACE = "\\[(.*?)\\]";
```
Comment (carledriss) [inline pivotal-api-test/src/main/java/com/fundacionjala/pivotalapi/LocalStore.java]: same as above

File: pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/Hook.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+import org.apache.log4j.Logger;
+
+/**
+ * Created by RosarioGarcia on 7/4/2016.
+ */
+public class Hook {
+
+
+    private static final Logger LOGGER = Logger.getLogger(PropertiesInfo.class);
+    public static final String BASE_PROJECTS = "/projects/";
```
Comment (carledriss) [inline pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/Hook.java]: This should be private

File: pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/Hook.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+import com.jayway.restassured.response.Response;
+import cucumber.api.java.After;
+import org.apache.log4j.Logger;
+
+/**
+ * Created by RosarioGarcia on 7/4/2016.
+ */
+public class Hook {
+
+
```
Comment (carledriss) [inline pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/Hook.java]: Remove this unnecessary line

File: pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/RequestStepDef.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * @author RosarioGarcia
+ */
+public class RequestStepDef {
+    private Response response;
+
+    @Given("^I send a POST request to (.*) with:$")
+    public void iSendAPOSTRequestToEndpoint(String endpoint, Map<String, Object> values) {
+        response = RequestManager.postRequest(LocalStore.formatEndpoint(endpoint), values);
```
Comment (carledriss) [inline pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/RequestStepDef.java]: Add import static for RequestManager and LocalStore

File: pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/RequestStepDef.java [hunk 20]
```diff
@@ -0,0 +1,40 @@
+package com.fundacionjala.pivotalapi.cucumber.stepdefinitions;
+
+import java.util.Map;
+
+import com.fundacionjala.pivotalapi.LocalStore;
```
Comment (carledriss) [inline pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/RequestStepDef.java]: Organize imports

File: pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/Hook.java [hunk 21]
```diff
@@ -0,0 +1,35 @@
+package com.fundacionjala.pivotalapi.cucumber.stepdefinitions;
+
+
+import com.fundacionjala.pivotalapi.PropertiesInfo;
```
Comment (carledriss) [inline pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/Hook.java]: Organize imports

File: pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/ResponseStepDef.java [hunk 22]
```diff
@@ -0,0 +1,29 @@
+package com.fundacionjala.pivotalapi.cucumber.stepdefinitions;
+
+import com.jayway.restassured.response.Response;
```
Comment (carledriss) [inline pivotal-api-test/src/test/java/com/fundacionjala/pivotalapi/cucumber/stepdefinitions/ResponseStepDef.java]: Organize imports

File: pivotal-api-test/target/test-classes/features/project/UpdateProject.feature [hunk 23]
```diff
@@ -0,0 +1,14 @@
+Feature: Update project settings
```
Comment (carledriss) [inline pivotal-api-test/target/test-classes/features/project/UpdateProject.feature:1]: This feature file and DeleteProject.feature have the same precondition (Background)
Please join them.

File: pivotal-api-test/target/test-classes/features/stories/EditStory.feature [hunk 24]
```diff
@@ -0,0 +1,19 @@
+Feature: Update a story in project from pivotal tracker
+
```
Comment (carledriss) [inline pivotal-api-test/target/test-classes/features/stories/EditStory.feature:2]: This feature file and RemoveStory.feature have the same precondition (Background)
Please join them.

**Comment by carledriss** [comment]
@rosariogarcia 
Don't forget to update this PR

**Comment by rosariogarcia** [comment]
@carledriss 
I update this PR

**Comment by carledriss** [comment]
Looks good to me!

**Pull request merged**